pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -256,20 +256,35 @@ func (s *DiscoveryServer) updateClusterInc(push *model.PushContext, clusterName
 	}
 	edsInstances.With(prometheus.Labels{"cluster": clusterName}).Set(float64(cnt))
 
-	// There is a chance multiple goroutines will update the cluster at the same time.
-	// This could be prevented by a lock - but because the update may be slow, it may be
-	// better to accept the extra computations.
-	// We still lock the access to the LoadAssignments.
-	edsCluster.mutex.Lock()
-	defer edsCluster.mutex.Unlock()
-
-	edsCluster.LoadAssignment = &xdsapi.ClusterLoadAssignment{
-		ClusterName: clusterName,
-		Endpoints:   locEps,
-	}
-	if len(locEps) > 0 && edsCluster.NonEmptyTime.IsZero() {
-		edsCluster.NonEmptyTime = time.Now()
+	for locality, edsCluster := range edsClusters {
+		// shallow copy locEps LocalityLbEndpoints,
+		// because below may mutate its LoadBalancingWeight or Priority
+		clonedLocEps := util.CloneLocalityLbEndpoints(locEps)
+
+		// There is a chance multiple goroutines will update the cluster at the same time.
+		// This could be prevented by a lock - but because the update may be slow, it may be
+		// better to accept the extra computations.
+		// We still lock the access to the LoadAssignments.
+		edsCluster.mutex.Lock()
```

Comment:
Have you tested this code with more than one locality ? 
I really can't see how it can work - it'll have a deadlock since mutex will be locked 2x.

Based on this - let's rollback the PR until it has more testing.... Can't imagine any scenario where this will not deadlock 
even users who don't enable the feature.

Reply:
Can you elaborate why this can cause a deadlock? And why ` mutex will be locked 2x`?

Reply:
you have 2 clusters. First loop iteration - edsCluster.mutex.Lock is called, plus a defer ( which will happen at the end of the method )
Second iteration - the mutex will block, it's not re-entrant. 

You need to do lock/unlock, can't use defer in a loop 
( we had this bug few times :-)

Reply:
I agree using defer in a loop is not good. But `you have 2 clusters. ` doesnot matter , because `edsCluster.mutex` is not same, its `EdsCluster` private.